craft_application/launchpad/launchpad.py

```diff
@@ -0,0 +1,138 @@
+# to complain a lot. As such, we're disabling several pyright checkers for this file
+# since in this case they generate more noise than utility.
+# pyright: reportUnknownMemberType=false
+
+from __future__ import annotations
+
+import pathlib
+
+import launchpadlib.launchpad  # type: ignore[import-untyped]
+import launchpadlib.uris  # type: ignore[import-untyped]
+import platformdirs
+from typing_extensions import Any, Self
+
+from . import models
+
+DEFAULT_CACHE_PATH = platformdirs.user_cache_path("launchpad-client")
+
+
+class Launchpad:
+    """A client for Launchpad."""
+
+    def __init__(
+        self, app_name: str, launchpad: launchpadlib.launchpad.Launchpad
+    ) -> None:
+        self.app_name = app_name
+        self.lp = launchpad
+
+    @classmethod
+    def anonymous(
+        cls,
+        app_name: str,
+        root: str = launchpadlib.uris.LPNET_SERVICE_ROOT,
+        cache_dir: pathlib.Path | None = DEFAULT_CACHE_PATH,
+        version: str = "devel",
+        timeout: int | None = None,
+    ) -> Self:
+        """Get an anonymous Launchpad client."""
+        return cls(
+            app_name,
+            launchpadlib.launchpad.Launchpad.login_anonymously(
+                consumer_name=app_name,
+                service_root=root,
+                launchpadlib_dir=cache_dir,
+                version=version,
+                timeout=timeout,
+            ),
+        )
+
+    @classmethod
+    def login(
+        cls,
+        app_name: str,
+        root: str = launchpadlib.uris.LPNET_SERVICE_ROOT,
+        cache_dir: pathlib.Path | None = DEFAULT_CACHE_PATH,
+        credentials_file: pathlib.Path | None = None,
+        version: str = "devel",
+        **kwargs: Any,  # noqa: ANN401 (Intentionally duck-typed)
+    ) -> Self:
+        """Login to Launchpad."""
+        return cls(
+            app_name,
+            launchpadlib.launchpad.Launchpad.login_with(
+                application_name=app_name,
+                service_root=root,
+                launchpadlib_dir=cache_dir,
+                credentials_file=credentials_file,
+                version=version,
+                **kwargs,
+            ),
+        )
+
+    def __repr__(self) -> str:
+        return f"{self.__class__.__name__}({self.app_name!r})"
+
+    def get_recipe(
+        self,
+        type_: models.RecipeType | str,
+        name: str,
+        owner: str,
+        project: str | None = None,
+    ) -> models.Recipe:
+        """Get a recipe.
+
+        :param type_: The type of recipe to get.
+        :param owner: The owner of the recipe.
+        :param name: The name of the recipe.
+        :param project: (Optional) The project to which the recipe is attached
+        :returns: A Recipe object of the appropriate type.
+        :raises: ValueError if the type requested is invalid.
+
+        A charm recipe requires a project. If none is given, the first charm matching
```

Comment:
should this bit of the docstring be made more generic?

Reply:
It's noting about a special case, but I'll make that more obvious.